File: app/src/main/java/org/tasks/repeats/BasicRecurrenceDialog.java
package org.tasks.repeats;

import static android.app.Activity.RESULT_OK;
import static com.google.common.collect.Lists.newArrayList;
import static net.fortuna.ical4j.model.Recur.Frequency.DAILY;
import static net.fortuna.ical4j.model.Recur.Frequency.HOURLY;
import static net.fortuna.ical4j.model.Recur.Frequency.MINUTELY;
import static net.fortuna.ical4j.model.Recur.Frequency.MONTHLY;
import static net.fortuna.ical4j.model.Recur.Frequency.WEEKLY;
import static net.fortuna.ical4j.model.Recur.Frequency.YEARLY;
import static org.tasks.Strings.isNullOrEmpty;
import static org.tasks.repeats.CustomRecurrenceDialog.newCustomRecurrenceDialog;
import static org.tasks.repeats.RecurrenceUtils.newRecur;
import static org.tasks.time.DateTimeUtils.currentTimeMillis;

import android.app.Activity;
import android.app.Dialog;
import android.content.Intent;
import android.os.Bundle;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.DialogFragment;
import androidx.fragment.app.Fragment;
import dagger.hilt.android.AndroidEntryPoint;
import java.util.List;
import javax.inject.Inject;
import net.fortuna.ical4j.model.Recur;
import net.fortuna.ical4j.model.Recur.Frequency;
import org.tasks.R;
import org.tasks.dialogs.DialogBuilder;
import org.tasks.ui.SingleCheckedArrayAdapter;
import timber.log.Timber;

@AndroidEntryPoint
public class BasicRecurrenceDialog extends DialogFragment {

  public static final String EXTRA_RRULE = "extra_rrule";
  private static final String EXTRA_DATE = "extra_date";
  private static final String FRAG_TAG_CUSTOM_RECURRENCE = "frag_tag_custom_recurrence";

  @Inject Activity context;
  @Inject DialogBuilder dialogBuilder;
  @Inject RepeatRuleToString repeatRuleToString;

  public static BasicRecurrenceDialog newBasicRecurrenceDialog(
      Fragment target, int rc, String rrule, long dueDate) {
    BasicRecurrenceDialog dialog = new BasicRecurrenceDialog();
    dialog.setTargetFragment(target, rc);
    Bundle arguments = new Bundle();
    if (rrule != null) {
      arguments.putString(EXTRA_RRULE, rrule);
    }
    arguments.putLong(EXTRA_DATE, dueDate);
    dialog.setArguments(arguments);
    return dialog;
  }

  @NonNull
  @Override
  public Dialog onCreateDialog(@Nullable Bundle savedInstanceState) {
    Bundle arguments = getArguments();
    long dueDate = arguments.getLong(EXTRA_DATE, currentTimeMillis());
    String rule = arguments.getString(EXTRA_RRULE);
    Recur rrule = null;
    try {
      if (!isNullOrEmpty(rule)) {
        rrule = newRecur(rule);
      }
    } catch (Exception e) {
      Timber.e(e);
    }

    boolean customPicked = isCustomValue(rrule);
    List<String> repeatOptions =
        newArrayList(context.getResources().getStringArray(R.array.repeat_options));
    SingleCheckedArrayAdapter adapter =
        new SingleCheckedArrayAdapter(context, repeatOptions);
    int selected = 0;
    if (customPicked) {
      adapter.insert(repeatRuleToString.toString(rule), 0);
    } else if (rrule != null) {
      switch (rrule.getFrequency()) {
        case DAILY:
          selected = 1;
          break;
        case WEEKLY:
          selected = 2;
          break;
        case MONTHLY:
          selected = 3;
          break;
        case YEARLY:
          selected = 4;
          break;
        default:
          selected = 0;
          break;
      }
    }
    return dialogBuilder
        .newDialog()
        .setSingleChoiceItems(
            adapter,
            selected,
            (dialogInterface, i) -> {
              if (customPicked) {
                if (i == 0) {
                  dialogInterface.dismiss();
                  return;
                }
                i--;
              }
              Recur result;
              if (i == 0) {
                result = null;
              } else if (i == 5) {
                newCustomRecurrenceDialog(
                    getTargetFragment(), getTargetRequestCode(), rule, dueDate)
                    .show(getParentFragmentManager(), FRAG_TAG_CUSTOM_RECURRENCE);
                dialogInterface.dismiss();
                return;
              } else {
                result = newRecur();
                result.setInterval(1);

                switch (i) {
                  case 1:
                    result.setFrequency(DAILY.name());
                    break;
                  case 2:
                    result.setFrequency(WEEKLY.name());
                    break;
                  case 3:
                    result.setFrequency(MONTHLY.name());
                    break;
                  case 4:
                    result.setFrequency(YEARLY.name());
                    break;
                }
              }

              Intent intent = new Intent();
              intent.putExtra(EXTRA_RRULE, result == null ? null : result.toString());
              getTargetFragment().onActivityResult(getTargetRequestCode(), RESULT_OK, intent);
              dialogInterface.dismiss();
            })
        .setOnCancelListener(null)
        .show();
  }

  private boolean isCustomValue(Recur rrule) {
    if (rrule == null) {
      return false;
    }
    Frequency frequency = rrule.getFrequency();
    return (frequency == WEEKLY || frequency == MONTHLY) && !rrule.getDayList().isEmpty()
        || frequency == HOURLY
        || frequency == MINUTELY
        || rrule.getUntil() != null
        || rrule.getInterval() > 1
        || rrule.getCount() > 0;
  }
}

File: app/src/main/java/org/tasks/repeats/BasicRecurrenceDialog.kt
package org.tasks.repeats

import android.app.Activity.RESULT_OK
import android.app.Dialog
import android.os.Bundle
import androidx.activity.result.contract.ActivityResultContracts
import androidx.core.os.bundleOf
import androidx.fragment.app.DialogFragment
import dagger.hilt.android.AndroidEntryPoint
import org.tasks.compose.pickers.BasicRecurrence
import org.tasks.compose.pickers.BasicRecurrenceOption
import org.tasks.compose.pickers.isCustomRecurrence
import org.tasks.dialogs.DialogBuilder
import org.tasks.repeats.CustomRecurrenceActivity.Companion.EXTRA_ACCOUNT_TYPE
import org.tasks.repeats.CustomRecurrenceActivity.Companion.newIntent
import org.tasks.repeats.RecurrenceUtils.newRecur
import org.tasks.themes.TasksTheme
import org.tasks.themes.Theme
import timber.log.Timber
import javax.inject.Inject

@AndroidEntryPoint
class BasicRecurrenceDialog : DialogFragment() {
    @Inject lateinit var dialogBuilder: DialogBuilder
    @Inject lateinit var repeatRuleToString: RepeatRuleToString
    @Inject lateinit var theme: Theme

    private val customRecurrence =
        registerForActivityResult(ActivityResultContracts.StartActivityForResult()) {
            if (it.resultCode == RESULT_OK) {
                setResult(it.data?.getStringExtra(EXTRA_RRULE))
            }
            dismiss()
        }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val args = requireArguments()
        val rule = args.getString(EXTRA_RRULE)
        val rrule =
            rule
                .takeIf { !it.isNullOrBlank() }
                ?.let {
                    try {
                        newRecur(it)
                    } catch (e: Exception) {
                        Timber.e(e)
                        null
                    }
                }
        val customPicked = rrule?.isCustomRecurrence() == true
        val customLabel = if (customPicked) repeatRuleToString.toStringBlocking(rule) else null
        val selectedFrequency = if (!customPicked) rrule?.frequency else null
        return dialogBuilder
            .newDialog()
            .setContent {
                TasksTheme(
                    theme = theme.themeBase.index,
                    primary = theme.themeColor.primaryColor,
                ) {
                    BasicRecurrence(
                        customLabel = customLabel,
                        selectedFrequency = selectedFrequency,
                        onSelected = { option -> onSelected(option, rule, args) },
                    )
                }
            }
            .show()
    }

    private fun onSelected(option: BasicRecurrenceOption, rule: String?, args: Bundle) {
        when (option) {
            BasicRecurrenceOption.KeepCustom -> dismiss()
            BasicRecurrenceOption.DoesNotRepeat -> {
                setResult(null)
                dismiss()
            }
            is BasicRecurrenceOption.Frequency -> {
                val recur = newRecur().apply {
                    interval = 1
                    setFrequency(option.frequency.name)
                }
                setResult(recur.toString())
                dismiss()
            }
            BasicRecurrenceOption.Custom -> {
                customRecurrence.launch(
                    newIntent(
                        context = requireContext(),
                        rrule = rule,
                        dueDate = args.getLong(EXTRA_DATE),
                        accountType = args.getInt(EXTRA_ACCOUNT_TYPE)
                    )
                )
            }
        }
    }

    private fun setResult(rrule: String?) {
        parentFragmentManager.setFragmentResult(
            REQUEST_KEY,
            bundleOf(EXTRA_RRULE to rrule)
        )
    }

    companion object {
        const val EXTRA_RRULE = "extra_rrule"
        const val REQUEST_KEY = "basic_recurrence_result"
        private const val EXTRA_DATE = "extra_date"

        fun newBasicRecurrenceDialog(
            rrule: String?,
            dueDate: Long,
            accountType: Int,
        ): BasicRecurrenceDialog {
            val dialog = BasicRecurrenceDialog()
            dialog.arguments = Bundle().apply {
                rrule?.let { putString(EXTRA_RRULE, it) }
                putLong(EXTRA_DATE, dueDate)
                putInt(EXTRA_ACCOUNT_TYPE, accountType)
            }
            return dialog
        }
    }
}
